File: python/src/siphon/providers/anthropic_provider.py
"""Anthropic provider adapter: beta Files API + Messages API content blocks."""

from __future__ import annotations

from typing import AsyncIterator, Iterator

from siphon.providers import ProviderRef

_SUPPORTED_PREFIXES = ("image/", "application/pdf", "text/")
_INLINE_SIZE_LIMIT = 5 * 1024 * 1024  # Anthropic's documented base64 image limit
_BETA_HEADER = ["files-api-2025-04-14"]


class AnthropicProvider:
    def __init__(self, client) -> None:
        self._client = client

    def supports_media_type(self, mime_type: str) -> bool:
        return mime_type.startswith(_SUPPORTED_PREFIXES)

    def inline_size_limit(self) -> int:
        return _INLINE_SIZE_LIMIT

    def upload(self, chunks: Iterator[bytes], mime_type: str) -> ProviderRef:
        data = b"".join(chunks)
        file_obj = self._client.beta.files.upload(
            file=("upload", data, mime_type),
            betas=_BETA_HEADER,
        )
        return ProviderRef(id=file_obj.id, expires_at=None)

    async def upload_async(
        self, chunks: AsyncIterator[bytes] | Iterator[bytes], mime_type: str
    ) -> ProviderRef:
        if hasattr(chunks, "__anext__"):
            data = b"".join([c async for c in chunks])  # type: ignore[union-attr]
        else:
            data = b"".join(chunks)  # type: ignore[arg-type]
        file_obj = await self._client.beta.files.upload(
            file=("upload", data, mime_type),
            betas=_BETA_HEADER,
        )
        return ProviderRef(id=file_obj.id, expires_at=None)

    def build_reference_block(self, ref: ProviderRef, mime_type: str) -> dict:
        block_type = "image" if mime_type.startswith("image/") else "document"
        return {"type": block_type, "source": {"type": "file", "file_id": ref.id}}

    def build_inline_block(self, base64_data: str, mime_type: str) -> dict:
        block_type = "image" if mime_type.startswith("image/") else "document"
        return {
            "type": block_type,
            "source": {"type": "base64", "media_type": mime_type, "data": base64_data},
        }

File: typescript/src/providers/anthropicProvider.ts
import type Anthropic from "@anthropic-ai/sdk";
import type { Provider, ProviderRef } from "./types.js";

const SUPPORTED_PREFIXES = ["image/", "application/pdf", "text/"];
const INLINE_SIZE_LIMIT = 5 * 1024 * 1024;
const BETA_HEADER = ["files-api-2025-04-14"];

export class AnthropicProvider implements Provider {
  constructor(private client: Anthropic) {}

  supportsMediaType(mimeType: string): boolean {
    return SUPPORTED_PREFIXES.some((prefix) => mimeType.startsWith(prefix));
  }

  inlineSizeLimit(): number {
    return INLINE_SIZE_LIMIT;
  }

  async upload(chunks: AsyncIterable<Buffer>, mimeType: string): Promise<ProviderRef> {
    const parts: Buffer[] = [];
    for await (const chunk of chunks) parts.push(chunk);
    const data = Buffer.concat(parts);

    const file = await this.client.beta.files.upload({
      file: new Blob([data], { type: mimeType }),
      betas: BETA_HEADER,
    });

    return { id: file.id, expiresAt: null };
  }

  buildReferenceBlock(ref: ProviderRef, mimeType: string): Record<string, unknown> {
    const type = mimeType.startsWith("image/") ? "image" : "document";
    return { type, source: { type: "file", file_id: ref.id } };
  }

  buildInlineBlock(base64Data: string, mimeType: string): Record<string, unknown> {
    const type = mimeType.startsWith("image/") ? "image" : "document";
    return { type, source: { type: "base64", media_type: mimeType, data: base64Data } };
  }
}
